examples/metalk8s-solution-example/manifest.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+
+"""Utility script to generate a Solution manifest from given arguments."""
+
+import argparse
+import sys
+import yaml
+
+
+SOLUTION_MANIFEST = {
+    'apiVersion': 'solutions.metalk8s.scality.com/v1alpha1',
+    'kind': 'Solution',
+    'metadata': {
+        'name': '',
+        'annotations': {},
+        'labels': {},
+    },
+    'spec': {
+        'images': [],
+        'operator': {
+            'image': {},
+        },
+        'ui': {
+            'image': {},
+        },
+        'version': '',
+    },
+}
+
+
+def parse_arguments():
+    """Parse arguments from the CLI."""
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        '-a',
+        '--annotation',
+        nargs=2,
+        action='append',
+        dest='annotations',
+        default=[],
+        metavar=('KEY', 'VALUE'),
+        help="add an annotation to the Solution manifest metadata",
+    )
+
+    parser.add_argument(
+        '-e',
+        '--extra-image',
+        nargs=2,
+        action='append',
+        dest='extra_images',
+        default=[],
+        metavar=('NAME', 'TAG'),
+        help="add an image to the list of images shipped by the Solution",
+    )
+
+    parser.add_argument(
+        '-l',
+        '--label',
+        nargs=2,
+        action='append',
+        dest='labels',
+        default=[],
+        metavar=('KEY', 'VALUE'),
+        help="add a label to the Solution manifest metadata",
+    )
+
+    parser.add_argument(
+        '-n',
+        '--name',
+        required=True,
+        metavar='SOLUTION_NAME',
+        help="set the name of the Solution",
+    )
+
+    parser.add_argument(
+        '-o',
+        '--operator-image',
+        required=True,
+        nargs=2,
+        metavar=('NAME', 'TAG'),
+        help="set the Operator image name and tag in the manifest spec",
+    )
+
+    parser.add_argument(
+        '-u',
+        '--ui-image',
+        required=True,
+        nargs=2,
+        metavar=('NAME', 'TAG'),
+        help="set the UI image name and tag in the manifest spec",
+    )
+
+    parser.add_argument(
+        '-v',
+        '--version',
+        required=True,
+        help="version of the Solution",
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    """Script entrypoint."""
+    args = parse_arguments()
+
+    manifest = SOLUTION_MANIFEST
```

Comment:
FYI: if it's something you're mutating in your method, then it can't be a constant. Hence, can't be using the constants naming conventions, to avoid confusing a user.

If the intention is to generate a new object from a set of args, then just define this object in the method body. This way, there's no risk of side effects:

```python
def _img_dict(name, tag):
    return {'image': {'name': name, 'tag': tag}}

def build_manifest(args):
    return {
        'apiVersion': 'solutions.metalk8s.scality.com/v1alpha1',
        'kind': 'Solution',
        'metadata': {
            'name': args.name,
            'annotations': dict(args.annotations),
            'labels': dict(args.labels),
        },
        'spec': {
            'images': list(set(
                ':'.join(image) for image in (
                    args.extra_images + [args.ui_image, args.operator_image]
                )
            )),
            'ui': _img_dict(*args.ui_image),
            'operator': _img_dict(*args.operator_image),
            'version': args.version,
        },
    }

# Try to make your functions testable
# Here, we can pass in a stream to output the manifest string,
# so in a test we can just pass an in-memory buffer and inspect it
def main(stream=sys.stdout):
    args = parse_arguments()
    manifest = build_manifest(args)

    # Also, no need to manually write,
    # just pass the stream to `yaml.safe_dump`
    yaml.safe_dump(manifest, stream, default_flow_style=False)
```

Reply:
It's just a helper script, so no need to change anything. But it's not longer to write code "the right way", so I hope you'll think about it in the future ;)

Reply:
Yes, I know this, but since it's just a simple script to generate a YAML, I though it would be sufficient.